# Committee Reference &ndash; Data Quality #

## Overview ##

Committees in the FEC data set have a unique ID assigned to them.  However, since we are combining Committee records from multiple election cycle source files, we really should join to `cmte` using both `cmte_id` and `elect_cycle`.  The purpose of this notebook is to explore the "quality" of the Committee data, both within and across election cycles, to see how consistent it is, and how we can create a unified Committee Master entity to improve referential integrity for the larger, complete data set.

Here is a list of the examinations in this notebook:

* Quality of Committee Names (`cmte_nm`)
* Integrity of Committee ID (`cmte_id`)
* Consistency of names for `cmte_id`'s across election cycles
* Multiple `cmte_id`'s for identical names &ndash; within election cycles
* Multiple `cmte_id`'s for identical names &ndash; across election cycles

Here are additional examinations to do:

* Multiple `cmte_id`'s for *similar* names &ndash; within election cycles
* Multiple `cmte_id`'s for *similar* names &ndash; across election cycles

## Notebook Setup ##

### Configure database connect info/options ###

Note: database connect string can be specified on the initial `%sql` command:

```python
database_url = "postgresql+psycopg2://user@localhost/fecdb"
%sql $database_url

```

Or, connect string is taken from DATABASE_URL environment variable (if not specified for `%sql`):

```python
%sql

```

In [1]:
%load_ext sql
%config SqlMagic.autopandas=True
%config InteractiveShell.ast_node_interactivity='last_expr_or_assign'
# connect string taken from DATABASE_URL environment variable
%sql

'Connected: crash@fecdb'

### Configure Python modules ###

In [2]:
import pandas as pd

pd.set_option("display.max_rows", 200)

### Set styling ###

In [3]:
%%html
<style>
  tr, th, td {
    text-align: left !important;
  }
</style>

## Examination of Committee Data Set ##

### High-level summary ###

First count total records and distinct `cmte_id`'s (and save out results for reference)

In [4]:
%%sql result <<
select count(*) as count_total,
       count(distinct cmte_id) as count_distinct_ids,
       count(distinct cmte_nm) as count_distinct_names
  from cmte

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.
Returning data to local variable result


In [5]:
cmte_count_total    = int(result.loc[0][0])
cmte_distinct_ids   = int(result.loc[0][1])
cmte_distinct_names = int(result.loc[0][2])
"cmte_count_total = %d, cmte_distinct_ids = %d, cmte_distinct_names = %d" % \
    (cmte_count_total, cmte_distinct_ids, cmte_distinct_names)

'cmte_count_total = 138384, cmte_distinct_ids = 45868, cmte_distinct_names = 50614'

### Quality of Committee Names (`cmte_nm`) ###

Let's try and get a sense of the extent of formatting problems (inconsistencies or flaws).  First look for names that have lowercase letters (uppercase is now the standard)...

In [6]:
%%sql
select elect_cycle,
       count(*)
  from cmte
 where cmte_nm ~ '[a-z]'
 group by 1
 order by 1

 * postgresql+psycopg2://crash@caladan/fecdb
7 rows affected.


elect_cycle  count
0         2004      4
1         2006      3
2         2008     11
3         2010      3
4         2012      2
5         2014      2
6         2016      1

In [7]:
%%sql
select cmte_nm,
       array_agg(distinct elect_cycle)
  from cmte
 where cmte_nm ~ '[a-z]'
 group by 1
 order by 1

 * postgresql+psycopg2://crash@caladan/fecdb
15 rows affected.


cmte_nm                       array_agg
0                         Catholics United        [2010, 2012, 2014, 2016]
1    Defenders of Wildlife Action Fund 527                          [2008]
2        Defenders of Willdife Action Fund                          [2008]
3                     Environment Colorado                          [2008]
4                   Environment New Jersey                          [2008]
5           Free Enterprise Fund Committee                          [2006]
6        League of Conservation Voters Inc                          [2008]
7                         Let Freedom Ring                          [2008]
8                 NARAL Pro-Choice America  [2004, 2008, 2010, 2012, 2014]
9      Planned Parenthood Action Fund Inc.        [2004, 2006, 2008, 2010]
10  Planned Parenthood Advocates Mar Monte                    [2004, 2006]
11  Planned Parenthood Pennsylvania Advoca                          [2008]
12                            PowerPac.org                          [2008]
13   The Planned Parenthood Action Fund of                          [2004]
14                   Wisconsin Environment                          [2008]

Next, look for names with consecutive whitespace...

In [8]:
%%sql
select elect_cycle,
       count(*)
  from cmte
 where cmte_nm ~ '\s{2,}'
 group by 1
 order by 1

 * postgresql+psycopg2://crash@caladan/fecdb
11 rows affected.


elect_cycle  count
0          2000      6
1          2002     33
2          2004     61
3          2006     84
4          2008     86
5          2010     80
6          2012     78
7          2014     68
8          2016     85
9          2018     90
10         2020     66

In [9]:
%%sql
select cmte_nm,
       array_agg(distinct elect_cycle)
  from cmte
 where cmte_nm ~ '\s{2,}'
 group by 1
 order by 1
 limit 50

 * postgresql+psycopg2://crash@caladan/fecdb
50 rows affected.


cmte_nm  \
0   ACADIAN AMBULANCE  SERVICE EMPLOYEE FEDERAL PO...   
1   ADVANCING CHIROPRACTIC PAC  (FEDERATION OF STR...   
2              ALASKA FEDERATION  OF REPUBLICAN WOMEN   
3                ALEXANDER AND  BALDWIN INC FEDPAC (A   
4   AMEGY BANK  NATIONAL  ASSOC. POLITICAL ACTION ...   
5   AMERICAN BEVERAGE ASSOCIATION POLITICAL ACTION...   
6      AMERICAN CLARITY AND EXCEPTIONALISM (ACE  PAC)   
7   AMERICAN FEDERATION OF STATE COUNTY & MUNICIPA...   
8   AMERICAN PODIATRIC MEDICAL ASSN. INC.  PODIATR...   
9                           AMERICAN RELIGIOUS  PARTY   
10  AM WATER WORKS CO INC ALLIANCE FOR AN INFORMED...   
11  AM. WATER WORKS CO. INC.  ALLIANCE FOR INFORME...   
12  AM. WATER WORKS CO., INC.  ALLIANCE FOR INFORM...   
13  APOLLO GROUP INC.  POLITICAL ORGANIZATION FOR ...   
14    APPLEPAC FKA:  APPLEGATE FOR CONGRESS COMMITTEE   
15  AQUILA INC EMPLOYEE FEDERAL PAC FORMERLY:  UTI...   
16  AQUILA INC. EMPLOYEE FEDERAL POLITICAL ACTION ...   
17    ARCHIPAC  -THE AMERICAN INSTITUTE OF ARCHITECTS   
18                                    AREVA  INC. PAC   
19  ATLANTIS ADVENTURES LLC POLITICAL  ACTION COMM...   
20  AUSTIN,  NICHOLS AND CO INCORPORATED/PERNOD RI...   
21  AVIVA USA  CORPORATION POLITICAL ACTION COMMITTEE   
22              AVIVA USA  POLITICAL ACTION COMMITTEE   
23  AZHHA  POLITICAL ACTION COMMITTEE-F/K/A PAC OF...   
24  BAACH ROBINSON  & LEWIS POLITICAL ACTION COMMI...   
25  BABCOCK  AND WILCOX COMPANY GOOD GOVERNMENT FU...   
26  BABCOCK & WILCOX COMPANY POLITICAL ACTION COMM...   
27  BAE SYSTEMS NORTH AMERICA INC.  POLITICAL ACTI...   
28  BAKER MANOCK & JENSEN  A PROFESSIONAL CORPORAT...   
29  BAYADA HOME HEALTH CARE, INC. POLITICAL ACTION...   
30  BETTER GOVERNMENT FUND OF MCDERMOTT INVESTMENT...   
31  BIPARTISAN VOLUNTARY PUBLIC AFFAIRS COMMITTEE ...   
32  BIPARTISAN VOLUNTARY PUBLIC AFFAIRS COMMITTEE ...   
33                                    BLANK  ROME PAC   
34            BLUE ACTION POLITICAL ACTION  COMMITTEE   
35               BOB AUERBACH FOR  CONGRESS COMMITTEE   
36               BOB  SAFFOLD  FOR CONGRESS COMMITTEE   
37                             BRENNER   FOR CONGRESS   
38   BRIGHTPOINT,  INC. D/B/A BRIGHTPOINT ECLIPSE PAC   
39                           BROADWATER  FOR CONGRESS   
40  BROWN  & WOOD LLP GOOD GOVERNMENT FUND (STATE ...   
41                             BRUDERLY  FOR CONGRESS   
42                       BUCHANAN FOR PRESIDENT  2004   
43  CA HOSPITAL ASSN PAC-FED  SPONSORED BY CA ASSN...   
44  CA HOSPITAL ASSN PAC-FED,  SPONSORED BY CA ASS...   
45  CA HOSPITAL ASSOCIATION PAC-FED SPONS BY CA AS...   
46  CALIFORNIA TEAMSTERS PUBLIC AFFAIRS COUNCIL,  ...   
47                CAPE FOX  PROFESSIONAL SERVICES LLC   
48  CAPITAL ONE FINANCIAL CORP. ASSOC. POLITICAL  ...   
49                                CARROLL  2 CONGRESS   

                                     array_agg  
0                                 [2018, 2020]  
1                     [2010, 2012, 2014, 2016]  
2                           [2014, 2016, 2018]  
3                           [2004, 2006, 2008]  
4                     [2006, 2008, 2010, 2012]  
5                                 [2006, 2008]  
6                                 [2018, 2020]  
7   [2008, 2010, 2012, 2014, 2016, 2018, 2020]  
8                                       [2006]  
9                                       [2020]  
10                                      [2002]  
11                          [2004, 2006, 2008]  
12                                      [2010]  
13              [2002, 2004, 2006, 2008, 2010]  
14                    [2006, 2008, 2010, 2012]  
15                                [2002, 2004]  
16                                [2006, 2008]  
17  [2006, 2008, 2010, 2014, 2016, 2018, 2020]  
18                    [2010, 2012, 2014, 2016]  
19                    [2006, 2008, 2010, 2012]  
20                          [2002, 2004, 2006]  
21                                [2010, 2012]  
22         

And now, names with errant spaces within parentheses...

In [10]:
%%sql
select elect_cycle,
       count(*)
  from cmte
 where cmte_nm ~ '\( | \)'
 group by 1
 order by 1

 * postgresql+psycopg2://crash@caladan/fecdb
11 rows affected.


elect_cycle  count
0          2000      1
1          2002      2
2          2004      1
3          2006      1
4          2008      3
5          2010      3
6          2012      6
7          2014      5
8          2016      7
9          2018      7
10         2020      5

In [11]:
%%sql
select cmte_nm,
       array_agg(distinct elect_cycle)
  from cmte
 where cmte_nm ~ '\( | \)'
 group by 1
 order by 1
 limit 50

 * postgresql+psycopg2://crash@caladan/fecdb
11 rows affected.


cmte_nm  \
0   BIPARTISAN VOLUNTARY PUBLIC AFFAIRS COMMITTEE ...   
1   CALIFORNIA FARM BUREAU FEDERATION FUND TO PROT...   
2   CALIFORNIA FARM BUREAU FUND TO PROTECT THE FAM...   
3   JOHN MANEELY COMPANY FAIR TRADE POLITICAL ACTI...   
4   MANTECH INTERNATIONAL CORPORATION POLITICAL AC...   
5   MANTECH INTERNATIONAL CORPORATION POLITICAL AC...   
6   SALT RIVER VALLEY WATER USERS' ASSOCIATION POL...   
7   SEIU CALIFORNIA STATE COUNCIL (NONPROFIT 501 (...   
8   STATE OF HAWAII ORGANIZATION OF POLICE OFFICER...   
9                  SYNTEX ( U S A ) LLC EMPLOYEES PAC   
10                     TRANSFORMING THE UNION ( T U )   

                                            array_agg  
0                                  [2012, 2014, 2016]  
1                      [2010, 2012, 2016, 2018, 2020]  
2                                              [2008]  
3                [2008, 2010, 2012, 2014, 2016, 2018]  
4                                  [2016, 2018, 2020]  
5                                        [2012, 2014]  
6   [2002, 2004, 2006, 2008, 2010, 2012, 2014, 201...  
7                                        [2018, 2020]  
8                      [2012, 2014, 2016, 2018, 2020]  
9                                        [2000, 2002]  
10                                       [2016, 2018]

And inconsistent spacing around commas (either space before, or no space after)

In [12]:
%%sql
select elect_cycle,
       count(*)
  from cmte
 where cmte_nm ~ ' ,|,[^ ]'
 group by 1
 order by 1

 * postgresql+psycopg2://crash@caladan/fecdb
11 rows affected.


elect_cycle  count
0          2000      3
1          2002      3
2          2004      6
3          2006      2
4          2008      2
5          2010      5
6          2012      4
7          2014      6
8          2016      6
9          2018      8
10         2020      8

In [13]:
%%sql
select cmte_nm,
       array_agg(distinct elect_cycle)
  from cmte
 where cmte_nm ~ ' ,|,[^ ]'
 group by 1
 order by 1
 limit 50

 * postgresql+psycopg2://crash@caladan/fecdb
22 rows affected.


cmte_nm  \
0                           10,000 LAKES VICTORY 2014   
1   BAKER, MANOCK & JENSEN , A PROFESSIONAL CORPOR...   
2                     BSA , THE SOFTWARE ALLIANCE PAC   
3                          CHAD LEE FOR CONGRESS,INC.   
4                         CITIZENS FOR MIKE ASSAD,INC   
5   CLOROX COMPANY EMPLOYEES' POLITICAL ACTION COM...   
6                             CONTI FOR CONGRESS,INC.   
7           DAVID WOOD , MAKE AMERICA RIGHTEOUS AGAIN   
8                                   DEMOCRACYONE,INC.   
9                DISTRICT 1199J,NUHHCE,AFSCME,AFL-CIO   
10                       FRIENDS FOR DARREN YANCY,INC   
11  I PEGGED DONALT TRUMP'S UGLY OLD BITCH ASS WIT...   
12                                        KAYDALY,COM   
13                                      LEININGER,O O   
14                              LIBERTY COMMITTEE,THE   
15  LOS ANGELES CONSTITUTIONAL COALITION OF BILL C...   
16  MEYER, SUOZZI, ENGLISH & KLEIN,PC-FEDERAL PAC(...   
17  PRO LIFE DEMOCRATS, ANGELS EXIST,THAY NEVER AS...   
18                                       SOROS,GEORGE   
19  STEVE SCHWARTZBERG FOR CONGRESS COMMITTEE,INCO...   
20  VOICE FOR TREES POLITICAL ACTION COMMITTEE, "V...   
21                WE CAN DO IT , YOU GUYS ! PROBABLY.   

                               array_agg  
0                                 [2014]  
1                                 [2010]  
2               [2014, 2016, 2018, 2020]  
3                           [2012, 2014]  
4                           [2014, 2016]  
5         [2000, 2002, 2004, 2006, 2008]  
6                                 [2004]  
7                                 [2020]  
8                                 [2020]  
9   [2010, 2012, 2014, 2016, 2018, 2020]  
10                                [2010]  
11                          [2018, 2020]  
12                    [2016, 2018, 2020]  
13              [2012, 2014, 2016, 2018]  
14  [2000, 2002, 2004, 2006, 2008, 2010]  
15                          [2010, 2012]  
16                    [2000, 2002, 2004]  
17                          [2018, 2020]  
18                                [2004]  
19                          [2018, 2020]  
20                                [2004]  
21                          [2016, 2018]

### Integrity of Committee ID (`cmte_id`) ###

Count records across election cycles and see if we have any null `cmte_id`'s (all zeros would be good)

In [14]:
%%sql
select elect_cycle,
       count(*)                  as records,
       count(*) - count(cmte_id) as null_cmte_ids
  from cmte
 group by 1
 order by 1

 * postgresql+psycopg2://crash@caladan/fecdb
11 rows affected.


elect_cycle  records  null_cmte_ids
0          2000     9577              0
1          2002     9103              0
2          2004     9322              0
3          2006     9282              0
4          2008    10017              0
5          2010    11138              0
6          2012    14455              0
7          2014    14905              0
8          2016    17827              0
9          2018    19270              0
10         2020    13488              0

Let's see if there are any duplicate `cmte_id`'s in any election cycles (should also be zero)

In [15]:
%%sql
with dup_cmte_id as (
    select elect_cycle,
           cmte_id,
           count(*) as id_count
      from cmte
     group by 1, 2
    having count(*) > 1
)
select elect_cycle,
       count(*) as dupes,
       sum(id_count) as total_dupe_ids,
       max(id_count) as max_dupe_ids
  from dup_cmte_id
 group by 1

 * postgresql+psycopg2://crash@caladan/fecdb
0 rows affected.


Empty DataFrame
Columns: []
Index: []

Now let's look at repeated `cmte_id`'s across election cycles (note that specifying `distinct` within `array_agg` is a tricky way of sorting the values, for consistency, if we care to group by that field)

In [16]:
%%sql
with cmte_id_sum as (
    select cmte_id,
           count(*) as ec_count,
           array_agg(distinct elect_cycle) as elect_cycles
      from cmte
     group by 1
)
select ec_count,
       count(*) as cmte_ids,
       round(count(*)::numeric / :cmte_distinct_ids * 100.0, 2) as pct_cmte_ids
  from cmte_id_sum
 group by 1
 order by 1 desc

 * postgresql+psycopg2://crash@caladan/fecdb
11 rows affected.


ec_count  cmte_ids pct_cmte_ids
0         11      2193         4.78
1         10       490         1.07
2          9       554         1.21
3          8       651         1.42
4          7       801         1.75
5          6      1338         2.92
6          5      2505         5.46
7          4      2971         6.48
8          3      5380        11.73
9          2     15998        34.88
10         1     12987        28.31

### Consistency of names for `cmte_id`'s across election cycles ###

Count the number of different names for Committee records (identified by `cmte_id`) appearing in multiple election cycles

In [17]:
%%sql
with cmte_diff_names as (
    select cmte_id,
           count(distinct cmte_nm)     as num_diff_names,
           array_agg(distinct cmte_nm) as diff_names
      from cmte
     group by 1
    having count(distinct cmte_nm) > 1
)
select count(*) as cmte_ids,
       round(count(*)::numeric / :cmte_distinct_ids * 100.0, 2) as pct_cmte_ids
  from cmte_diff_names

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


cmte_ids pct_cmte_ids
0      4572         9.97

Let's report by the different levels of variation on name (i.e. number of different representations)

In [18]:
%%sql
with cmte_diff_names as (
    select cmte_id,
           count(distinct cmte_nm)     as num_diff_names,
           array_agg(distinct cmte_nm) as diff_names
      from cmte
     group by 1
    having count(distinct cmte_nm) > 1
)
select num_diff_names,
       count(*) as cmte_ids,
       round(count(*)::numeric / :cmte_distinct_ids * 100.0, 2) as pct_cmte_ids
  from cmte_diff_names
 group by 1
 order by 1 desc

 * postgresql+psycopg2://crash@caladan/fecdb
8 rows affected.


num_diff_names  cmte_ids pct_cmte_ids
0               9         2         0.00
1               8         1         0.00
2               7         6         0.01
3               6        17         0.04
4               5        75         0.16
5               4       291         0.63
6               3       979         2.13
7               2      3201         6.98

Get an idea of what the different names associated with the same `cmte_id` look like&mdash;let's start with a sampling of Committee IDs with `num_diff_names` = 2 (compare adjacent `cmte_nm`'s)...

In [19]:
%%sql
with cmte_diff_names as (
    select cmte_id,
           count(distinct cmte_nm)     as num_diff_names,
           array_agg(distinct cmte_nm) as diff_names
      from cmte
     group by 1
    having count(distinct cmte_nm) = 2
)
select cmte_id,
       unnest(diff_names) as cmte_nm
  from cmte_diff_names
 order by cmte_id
 limit 50

 * postgresql+psycopg2://crash@caladan/fecdb
50 rows affected.


cmte_id                                            cmte_nm
0   C00000042  ILLINOIS TOOL WORKS FOR BETTER GOVERNMENT COMM...
1   C00000042  ILLINOIS TOOL WORKS INC. FOR BETTER GOVERNMENT...
2   C00000059                                 HALLMARK CARDS PAC
3   C00000059  HALLMARK POLITICAL ACTION COMMITTEE-FEDERAL HA...
4   C00000372                MAINTENANCE OF WAY POLITICAL LEAGUE
5   C00000372  MAINTENANCE OF WAY POLITICAL LEAGUE A PAC OF T...
6   C00000794                               LENT  & SCRIVNER PAC
7   C00000794                                LENT & SCRIVNER PAC
8   C00000901  BUILD POLITICAL ACTION COMMITTEE OF THE NATION...
9   C00000901  BUILD POLITICAL ACTION COMMITTEE OF THE NATION...
10  C00000992                 BRANCH COUNTY REPUBLICAN COMMITTEE
11  C00000992                          BRANCH COUNTY REPUBLICANS
12  C00001016  CARPENTERS LEGISLATIVE IMPROVEMENT COMMITTEE U...
13  C00001016  CARPENTERS LEGISLATIVE IMPROVEMENT COMM, UNITE...
14  C00001180  MICHIGAN DOCTORS POLITICAL ACTION COMMITTEE - ...
15  C00001180  MICHIGAN DOCTORS' POLITICAL ACTION COMMITTEE -...
16  C00001214  TEXAS MEDICAL ASSOCIATION POLICIAL ACTION COMM...
17  C00001214  TEXAS MEDICAL ASSOCIATION POLITICAL ACTION COM...
18  C00001313                      REPUBLICAN PARTY OF MINNESOTA
19  C00001313            REPUBLICAN PARTY OF MINNESOTA - FEDERAL
20  C00001347  HAWAII MEDICAL ASSOCIATION POLITICAL ACTION CO...
21  C00001347          HAWAII MEDICAL POLITICAL ACTION COMMITTEE
22  C00001461          ALASKA MEDICAL POLITICAL ACTION COMMITTEE
23  C00001461  ALASKA STATE MEDICAL ASSOCIATION POLITICAL ACT...
24  C00001727       BUSINESS INDUSTRY POLITICAL ACTION COMMITTEE
25  C00001727       BUSINESS-INDUSTRY POLITICAL ACTION COMMITTEE
26  C00001743  LOCAL 1814 ILA AFL-CIO POLITICAL ACTION AND ED...
27  C00001743  LOCAL 1814 INTERNATIONAL LONGSHOREMENS ASSOC A...
28  C00001818      WASHINGTON MEDICAL POLITICAL ACTION COMMITTEE
29  C00001818  WASHINGTON STATE MEDICAL ASSN POLITICAL ACTION...
30  C00001875  AMERICA'S COMMUNITY BANKERS COMMUNITY CAMPAIGN...
31  C00001875  AMERICA'S COMMUNITY BANKERS COMMUNITY CAMPAIGN...
32  C00002089  COMMUNICATIONS WORKERS OF AMERICA-COPE POLITIC...
33  C00002089         CWA-COPE POLITICAL CONTRIBUTIONS COMMITTEE
34  C00002238  NATIONAL COUNCIL OF FARMER COOPERATIVES CO-OP/PAC
35  C00002238  NATIONAL COUNCIL OF FARMER COOPERATIVES POLITI...
36  C00002261               FEDERATION OF AMERICAN HOSPITALS PAC
37  C00002261  FEDERATION OF AMERICAN HOSPITALS PAC (FEDPAC);...
38  C00002592                    COMMITTEE TO RE-ELECT TOM FOLEY
39  C00002592                     COMMITTEE TO REELECT TOM FOLEY
40  C00003061  NORTH DAKOTA COMMITTEE ON MEDICAL POLITICAL AC...
41  C00003061  NORTH DAKOTA MEDICAL ASSOCIATION POLITICAL ACT...
42  C00003152  NORTH CAROLINA MEDICAL SOCIETY FEDERAL POLITIC...
43  C00003152  NORTH CAROLINA MEDICAL SOCIETY POLITICAL EDUCA...
44  C00003210                                   UTAH MEDICAL PAC
45  C00003210            UTAH MEDICAL POLITICAL ACTION COMMITTEE
46  C00003251         NEA FUND FOR CHILDREN AND PUBLIC EDUCATION
47  C00003251  NEA FUND FOR CHILDREN AND PUBLIC EDUCATION; TH...
48  C00003327    OHIO MEDICAL POLITICAL ACTION COMMITTEE (OMPAC)
49  C00003327  OHIO STATE MEDICAL ASSOCIATION POLITICAL ACTIO...

Now, let's look at `num_diff_names` = 3...

In [20]:
%%sql
with cmte_diff_names as (
    select cmte_id,
           count(distinct cmte_nm)     as num_diff_names,
           array_agg(distinct cmte_nm) as diff_names
      from cmte
     group by 1
    having count(distinct cmte_nm) = 3
)
select cmte_id,
       unnest(diff_names) as cmte_nm
  from cmte_diff_names
 order by cmte_id
 limit 51

 * postgresql+psycopg2://crash@caladan/fecdb
51 rows affected.


cmte_id                                            cmte_nm
0   C00000489               D R I V E POLITICAL FUND CHAPTER 886
1   C00000489              D R I V E POLITICAL FUND, CHAPTER 886
2   C00000489  D R I V E POLITICAL FUND, TEAMSTERS LOCAL UNIO...
3   C00000638         INDIANA MEDICAL POLITICAL ACTION COMMITTEE
4   C00000638  INDIANA STATE MEDICAL ASSOCIATION POLITICAL AC...
5   C00000638  INDIANA STATE MEDICAL ASSOCIATION POLITICAL AC...
6   C00000729  AMERICAN DENTAL ASSOCIATION POLITICAL ACTION C...
7   C00000729             AMERICAN DENTAL POLITICAL ACTION CMTE.
8   C00000729         AMERICAN DENTAL POLITICAL ACTION COMMITTEE
9   C00000935                                               DCCC
10  C00000935        DEMOCRATIC CONGRESSIONAL CAMPAIGN COMMITTEE
11  C00000935  DEMOCRATIC CONGRESSIONAL CAMPAIGN COMMITTEE - ...
12  C00000984  UNITED STATES TELECOM ASSOCIATION POLITICAL AC...
13  C00000984  UNITED STATES TELECOM ASSOCIATION POLITICAL AC...
14  C00000984  UNITED STATES TELECOM ASSOCIATION POLITICAL AC...
15  C00001008       LOUISIANA MEDICAL POLITICAL ACTION COMMITTEE
16  C00001008  LOUISIANA STATE MEDICAL SOCIETY POLITICAL ACTI...
17  C00001008  LOUISIANA STATE MEDICAL SOCIETY POLITICAL ACTI...
18  C00001388  DAIRY FARMERS OF AMERICA, INC. - DEPAC (DAIRY ...
19  C00001388  DAIRY FARMERS OF AMERICA INC DEPAC (DAIRY EDUC...
20  C00001388  DAIRY FARMERS OF AMERICA, INC. POLITICAL ACTIO...
21  C00001636  INTERNATIONAL ASSOCIATION OF SHEET METAL, AIR,...
22  C00001636  UNITED TRANSPORTATION UNION POLITICAL ACTION C...
23  C00001636  UNITED TRANSPORTATION UNION (UTU) TRANSPORTATI...
24  C00001917                  WY DEMOCRATIC  STATE CENTRAL COMM
25  C00001917              WY DEMOCRATIC STATE CENTRAL COMMITTEE
26  C00001917         WYOMING STATE DEMOCRATIC CENTRAL COMMITTEE
27  C00002600                       JOHN D. DINGELL FOR CONGRESS
28  C00002600              JOHN D DINGELL FOR CONGRESS COMMITTEE
29  C00002600             JOHN D. DINGELL FOR CONGRESS COMMITTEE
30  C00002766  UNITED FOOD AND COMMERCIAL WORKERS INTERNATION...
31  C00002766  UNITED FOOD & COMMERCIAL WORKERS, ACTIVE BALLO...
32  C00002766  UNITED FOOD & COMMERCIAL WORKERS INTERNATIONAL...
33  C00002881  TRUCKING POLITICAL ACTION COMMITTEE OF THE AME...
34  C00002881  TRUCKING POLITICAL ACTION COMMITTEE OF THE AME...
35  C00002881  TRUCKING POLITICAL ACTION COMMITTEE OF THE AME...
36  C00003194  CALIFORNIA MEDICAL ASSOCIATION POLITICAL ACTIO...
37  C00003194      CALIFORNIA MEDICAL POLITICAL ACTION COMMITTEE
38  C00003194  CALIFORNIA MEDICAL POLITICAL ACTION COMMITTEE ...
39  C00003764                NATIONAL RESTAURANT ASSOCIATION PAC
40  C00003764  NATIONAL RESTAURANT ASSOCIATION PAC (RESTAURAN...
41  C00003764  NATIONAL RESTAURANT ASSOCIATION POLITICAL ACTI...
42  C00004275                                        ABA BANKPAC
43  C00004275               AMERICAN BANKERS ASSOCIATION BANKPAC
44  C00004275         AMERICAN BANKERS ASSOCIATION PAC (BANKPAC)
45  C00004473  NATIONAL TELECOMMUNICATIONS COOPERATIVE ASSOCI...
46  C00004473  NATIONAL TELEPHONE COOPERATIVE ASSOCIATION TEL...
47  C00004473  NAT'L TELECOMMUNICATIONS COOPERATIVE ASSN./TEL...
48  C00004499  ARIZONA MEDICAL ASSOCIATION INC POLICAL ACTION...
49  C00004499  ARIZONA MEDICAL ASSOCIATION, INC POLICAL ACTIO...
50  C00004499  ARIZONA MEDICAL ASSOCIATION, INC POLITICAL ACT...

And `num_diff_names` = 4...

In [21]:
%%sql
with cmte_diff_names as (
    select cmte_id,
           count(distinct cmte_nm)     as num_diff_names,
           array_agg(distinct cmte_nm) as diff_names
      from cmte
     group by 1
    having count(distinct cmte_nm) = 4
)
select cmte_id,
       unnest(diff_names) as cmte_nm
  from cmte_diff_names
 order by cmte_id
 limit 52

 * postgresql+psycopg2://crash@caladan/fecdb
52 rows affected.


cmte_id                                            cmte_nm
0   C00000885  INTERNATIONAL UNION OF PAINTERS & ALLIED TRADE...
1   C00000885  INTERNATIONAL UNION OF PAINTERS AND ALLIED TRA...
2   C00000885  INTERNATIONAL UNION OF PAINTERS AND ALLIED TRA...
3   C00000885  INTERNATIONAL UNION OF PAINTERS AND ALLIED TRA...
4   C00001198         AMERICAN HOTEL AND LODGING ASSOCIATION PAC
5   C00001198  AMERICAN HOTEL AND LODGING ASSOCIATION POLITIC...
6   C00001198  AMERICAN HOTEL & LODGING ASSOCIATION PAC FKA A...
7   C00001198    AMERICAN HOTEL MOTEL POLITICAL ACTION COMMITTEE
8   C00002469           MACHINISTS NON PARTISAN POLITICAL LEAGUE
9   C00002469           MACHINISTS NON-PARTISAN POLITICAL LEAGUE
10  C00002469  MACHINISTS NON PARTISAN POLITICAL LEAGUE OF TH...
11  C00002469  MACHINISTS NON PARTISAN POL LEAGUE OF THE INT'...
12  C00002972  ACTION COMMITTEE FOR RURAL ELECTRIFICATION (AC...
13  C00002972  ACTION COMMITTEE FOR RURAL ELECTRIFICATION. (A...
14  C00002972  ACTION COMMITTEE. FOR RURAL ELECTRIFICATION. (...
15  C00002972  THE NATIONAL RURAL ELECTRIC COOPERATIVE ASSOCI...
16  C00003160  NORTH AMERICA'S BUILDING TRADES UNIONS POLITIC...
17  C00003160  POLITICAL EDUCATIONAL FUND OF THE BUILDING AND...
18  C00003160  POLITICAL EDUCATIONAL FUND OF THE BUILDING AND...
19  C00003160  POLITICAL EDUCATIONAL FUND OF THE BUILDING AND...
20  C00003863  DISTRICT NO. 1-PCD, MARINE ENGINEERS' BENEFICI...
21  C00003863  DISTRICT NO. 1-PCD, MARINE ENGINEERS' BENEFICI...
22  C00003863  MARINE ENGINEERS BENEFICIAL ASOCIATION - RETIR...
23  C00003863  MARINE ENGINEERS BENEFICIAL ASSOCIATION RETIRE...
24  C00004036  SEIU COPE (SERVICE EMPLOYEES INTERNATIONAL UNI...
25  C00004036  SERVICE EMPLOYEES INTERNATIONAL UNION COMMITTE...
26  C00004036  SERVICE EMPLOYEES INTERNATIONAL UNION POLITICA...
27  C00004036  SERVICE EMPLOYEES INTERNATIONAL UNION POLITICA...
28  C00004325  SEAFARERS POLITICAL ACTIVITY DONATION - SEAFAR...
29  C00004325  SEAFARERS POLITICAL ACTIVITY DONATION - SEAFAR...
30  C00004325  SEAFARERS POLITICAL ACTIVITY DONATION-SEAFARER...
31  C00004325  SEAFARERS POLITICAL ACTIVITY DONATION-SEAFARER...
32  C00007898  OFFICE AND PROFESSIONAL EMPLOYEES INTERNATIONA...
33  C00007898  OFFICE AND PROFESSIONAL EMPLOYEES INTERNATIONA...
34  C00007898  OFFICE AND PROFESSIONAL EMPLOYEES INTERNATIONA...
35  C00007898       OPEIU JB MOSS VOICE OF THE ELECTORATE (VOTE)
36  C00009357  RECORDING INDUSTRY ASSOC/AMERICA INC POLITICAL...
37  C00009357  RECORDING INDUSTRY ASSOCIATION OF AMERICA POLI...
38  C00009357  THE RECORDING INDUSTRY ASSOCIATION OF AMERICA,...
39  C00009357  THE RECORDING INDUSTRY ASSOCIATION OF AMERICA ...
40  C00009639                SUNTRUST BANK GOOD GOVERNMENT GROUP
41  C00009639      SUNTRUST BANK GOOD GOVERNMENT GROUP - GEORGIA
42  C00009639  SUNTRUST BANKS INC. MAIN STREET BANKERS POLITI...
43  C00009639                   SUNTRUST MAIN STREET BANKERS PAC
44  C00012468  COCA-COLA COMPANY NONPARTISAN COMMITTEE FOR GO...
45  C00012468  COCA-COLA COMPANY NONPARTISAN COMMITTEE FOR GO...
46  C00012468  COCA-COLA COMPANY NONPARTISAN COMMITTEE FOR GO...
47  C00012468  THE COCA-COLA COMPANY NONPARTISAN COMMITTEE FO...
48  C00012476  UA UNION PLUMBERS & PIPEFITTERS VOTE! PAC (UNI...
49  C00012476   UNITED ASSOCIATION POLITICAL EDUCATION COMMITTEE
50  C00012476  UNITED ASSOCIATION POLITICAL EDUCATION COMMITT...
51  C00012476  UNITED ASSOCIATION POLITICAL EDUCATION COMMITT...

And for the fun of it, let's look at the most extreme examples (`num_diff_names` > 7)

In [22]:
%%sql
with cmte_diff_names as (
    select cmte_id,
           count(distinct cmte_nm)     as num_diff_names,
           array_agg(distinct cmte_nm) as diff_names
      from cmte
     group by 1
    having count(distinct cmte_nm) > 7
)
select cmte_id,
       unnest(diff_names) as cmte_nm
  from cmte_diff_names
 order by cmte_id
 limit 100

 * postgresql+psycopg2://crash@caladan/fecdb
26 rows affected.


cmte_id                                            cmte_nm
0   C00077701  KRAFT FOOD INC POLITICAL ACTION COMMITTEE (KF ...
1   C00077701  KRAFT FOODS GLOBAL INC. POLITICAL ACTION COMMI...
2   C00077701  KRAFT FOODS GLOBAL INC. POLITICAL ACTION COMMI...
3   C00077701  KRAFT FOODS GLOBAL INC. POLITICAL ACTION COMMI...
4   C00077701  KRAFT FOODS GROUP INC. POLITICAL ACTION COMMIT...
5   C00077701  KRAFT FOODS GROUP, INC. POLITICAL ACTION COMMI...
6   C00077701  KRAFT FOODS NORTH AMERICA INC. POLTICAL ACTION...
7   C00077701  THE KRAFT HEINZ COMPANY POLITICAL ACTION COMMI...
8   C00113753                          JOHNSON CONTROLS INC. PAC
9   C00113753       MALLINCKRODT INC. POLITICAL ACTION COMMITTEE
10  C00113753  MALLINCKRODT INC POLITICAL ACTION COMMITTEE (M...
11  C00113753  TYCO INTERNATIONAL MANAGEMENT COMPANY EMPLOYEE...
12  C00113753  TYCO INTERNATIONAL MANAGEMENT COMPANY EMPLOYEE...
13  C00113753          TYCO INTERNATIONAL MANAGEMENT COMPANY PAC
14  C00113753  TYCO INTERNATIONAL MANAGEMENT COMPANY PAC (TYC...
15  C00113753  TYCO INTERNATIONAL (US) INC. EMPLOYEES POLITIC...
16  C00113753  TYCO INTERNATIONAL (US) INC. EMPLOYEES POLITIC...
17  C00140590                        CALIFORNIA REPUBLICAN PARTY
18  C00140590           CALIFORNIA REPUBLICAN PARTY FEDERAL ACCT
19  C00140590          CALIFORNIA REPUBLICAN PARTY FEDERAL ACCT.
20  C00140590      CALIFORNIA REPUBLICAN PARTY / TEAM CALIFORNIA
21  C00140590                    CALIFORNIA REPUBLICAN PARTY/V10
22  C00140590                     CALIFORNIA REPUBLICAN PARTY/V8
23  C00140590           CALIFORNIA REPUBLICAN PARTY/VICTORY 2006
24  C00140590                  CALIFORNIA STATE REPUBLICAN PARTY
25  C00140590           VICTORY 2004/CALIFORNIA REPUBLICAN PARTY

### Multiple `cmte_id`'s for identical names &ndash; within election cycles ###

Let's first see how many names are involved, and what percentage of total distinct names that represents

In [23]:
%%sql
with shared_cmte_name as (
    select elect_cycle,
           cmte_nm,
           count(*) as num_shares
      from cmte
     where cmte_nm is not null
     group by 1, 2
    having count(*) > 1
)
select count(*) as shared_names,
       round(count(*)::numeric / :cmte_distinct_names * 100.0, 2) as pct_distinct_names
  from shared_cmte_name

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


shared_names pct_distinct_names
0          1412               2.79

Now we'll report by the level of replication (name sharing by different Committees, as identified by `cmte_id`) we have in various cycles; note that the same name may "offend" (map to multiple `cmte_id`'s) within different election cycles (either in the same, or differing, "num_shares")

In [24]:
%%sql
with shared_cmte_name as (
    select elect_cycle,
           cmte_nm,
           count(*) as num_shares
      from cmte
     where cmte_nm is not null
     group by 1, 2
    having count(*) > 1
)
select num_shares,
       count(*) as shared_names,
       array_agg(distinct elect_cycle) as elect_cycles
  from shared_cmte_name
 group by 1
 order by 1 desc

 * postgresql+psycopg2://crash@caladan/fecdb
4 rows affected.


num_shares  shared_names                                       elect_cycles
0           5             5                     [2000, 2012, 2016, 2018, 2020]
1           4            12         [2000, 2008, 2010, 2012, 2014, 2016, 2018]
2           3            70  [2000, 2002, 2004, 2006, 2008, 2010, 2012, 201...
3           2          1325  [2000, 2002, 2004, 2006, 2008, 2010, 2012, 201...

Let's take a look at some of the top offenders (see "elect_cycles" for multiple offenses of the same "num_shares" by a name)

In [25]:
%%sql
with shared_cmte_name as (
    select elect_cycle,
           cmte_nm,
           count(*) as num_shares,
           array_agg(distinct cmte_id) as cmte_ids
      from cmte
     where cmte_nm is not null
     group by 1, 2
    having count(*) > 2
)
select cmte_nm,
       array_length(cmte_ids, 1) num_cmte_ids,
       cmte_ids,
       array_agg(elect_cycle) as elect_cycles
  from shared_cmte_name
 group by 1, 3
 order by array_length(cmte_ids, 1) desc, count(*) desc, cmte_nm

 * postgresql+psycopg2://crash@caladan/fecdb
61 rows affected.


cmte_nm  num_cmte_ids  \
0                                COLLINS FOR CONGRESS             5   
1                                COLLINS FOR CONGRESS             5   
2                                  MOORE FOR CONGRESS             5   
3                          CLEAN WATER ACTION PROJECT             4   
4                                COLLINS FOR CONGRESS             4   
5                                COLLINS FOR CONGRESS             4   
6                                  KELLY FOR CONGRESS             4   
7                                  KELLY FOR CONGRESS             4   
8                                KENNEDY FOR CONGRESS             4   
9                                  MOORE FOR CONGRESS             4   
10                                PARKER FOR CONGRESS             4   
11                              PETERSON FOR CONGRESS             4   
12                              ROBINSON FOR CONGRESS             4   
13                              WILLIAMS FOR CONGRESS             4   
14                           SUSAN B ANTHONY LIST INC             3   
15                        AMERICAN CONSERVATIVE UNION             3   
16                           AMERICANS FOR PROSPERITY             3   
17                               COLLINS FOR CONGRESS             3   
18                    COMMITTEE TO ELECT GARY WINTERS             3   
19                                 LATINOS FOR REFORM             3   
20              NATIONAL ASSOCIATION OF HOME BUILDERS             3   
21  NATIONAL COMMITTEE TO PRESERVE SOCIAL SECURITY...             3   
22                     NATIONAL EDUCATION ASSOCIATION             3   
23                 NEVADA STATE EDUCATION ASSOCIATION             3   
24                            RICHARDSON FOR CONGRESS             3   
25                             SCHNEIDER FOR CONGRESS             3   
26           SCHRINER PRESIDENTIAL ELECTION COMMITTEE             3   
27                                        SIERRA CLUB             3   
28                                THOMAS FOR CONGRESS             3   
29                             US CHAMBER OF COMMERCE             3   
30                              ANDERSON FOR CONGRESS             3   
31                                 BROWN FOR CONGRESS             3   
32                              CAMPBELL FOR CONGRESS             3   
33                              CAMPBELL FOR CONGRESS             3   
34                                CANNON FOR CONGRESS             3   
35                                 CLARK FOR CONGRESS             3   
36                                 DAVIS FOR CONGRESS             3   
37                              FLETCHER FOR CONGRESS             3   
38                          ISAKSON VICTORY COMMITTEE             3   
39                                 KELLY FOR CONGRESS             3   
40                               KENNEDY FOR CONGRESS             3   
41                                 LUCAS FOR CONGRESS             3   
42                                     MILLENNIAL PAC             3   
43                                 MOORE FOR CONGRESS             3   
44                                 MOORE FOR CONGRESS             3   
45                                 MOORE FOR CONGRESS             3   
46                                MORGAN FOR CONGRESS             3   
47                                MORGAN FOR CONGRESS             3   
48                                PHELPS FOR CONGRESS             3   
49                                PORTER FOR CONGRESS             3   
50                               ROBERTS FOR CONGRESS             3   
51                              ROBINSON FOR CONGRESS             3   
52                                ROGERS FOR CONGRESS             3   
53                                ROGERS FOR CONGRESS             3   
54                               RUSSELL FOR CONGRESS             3   
55                                 SMITH FOR CONGRESS             3   

And we'll look at the top recurrences for cases where the number of shares is exactly two in an election cycle

In [26]:
%%sql
with shared_cmte_name as (
    select elect_cycle,
           cmte_nm,
           count(*) as num_shares,
           array_agg(distinct cmte_id) as cmte_ids
      from cmte
     where cmte_nm is not null
     group by 1, 2
    having count(*) = 2
)
select cmte_nm,
       cmte_ids,
       count(*) as num_elect_cycles,
       array_agg(elect_cycle) as elect_cycles
  from shared_cmte_name
 group by 1, 2
 order by 3 desc, 1, 2
 limit 50

 * postgresql+psycopg2://crash@caladan/fecdb
50 rows affected.


cmte_nm                cmte_ids  \
0                          FRELINGHUYSEN FOR CONGRESS  [C00148684, C00299404]   
1                          FOCUS ON THE FAMILY ACTION  [C30000673, C90008186]   
2                   LEAGUE OF CONSERVATION VOTERS INC  [C70004262, C90005786]   
3                             COMMON SENSE ISSUES INC  [C30001457, C90009739]   
4                                 GRAVES FOR CONGRESS  [C00359034, C00462556]   
5                               HUMAN RIGHTS CAMPAIGN  [C70004569, C90012626]   
6                                  PRICE FOR CONGRESS  [C00195628, C00386755]   
7        REFORM PARTY OF THE UNITED STATES OF AMERICA  [C00331314, C00364307]   
8                            VOTEVETS.ORG ACTION FUND  [C30001275, C90010620]   
9      AFL-CIO COPE POLITICAL CONTRIBUTIONS COMMITTEE  [C00003806, C70000112]   
10                               AMERICAN FUTURE FUND  [C30001028, C90011677]   
11                         AMERICANS FOR JOB SECURITY  [C30001135, C90011669]   
12                                           BILL PAC  [C00412288, C00527275]   
13                                 BLUEGREEN ALLIANCE  [C30001739, C90011743]   
14           CALIFORNIA LEAGUE OF CONSERVATION VOTERS  [C00012401, C90013103]   
15  CALIFORNIA NURSES ASSOCIATION/NATIONAL NURSES ...  [C30000814, C70004791]   
16      CALIFORNIA STATE COUNCIL OF SERVICE EMPLOYEES  [C70003199, C90012287]   
17            CROSSROADS GRASSROOTS POLICY STRATEGIES  [C30001655, C90011719]   
18                ENTERPRISE FREEDOM ACTION COMMITTEE  [C30001846, C90012162]   
19                  HILLSDALE COUNTY DEMOCRATIC PARTY  [C00355347, C90011180]   
20                          INDEPENDENT WOMEN'S VOICE  [C30001572, C90011115]   
21     INHUMANITY CHILD ABUSE CHRISTIANITY CAPITALISM  [C00386631, C00425744]   
22        NATIONAL FEDERATION OF INDEPENDENT BUSINESS  [C30001317, C70001243]   
23  NATIONAL RIFLE ASSOCIATION INSTITUTE FOR LEGIS...  [C70000716, C90013301]   
24                               RIGHTCHANGE.COM, INC  [C30001234, C90011248]   
25                               SEALASKA CORPORATION  [C70005384, C90012212]   
26    SERVICE EMPLOYEE INTERNATIONAL UNION LOCAL 1000  [C70005392, C90012634]   
27                            THE 60 PLUS ASSOCIATION  [C30001671, C90011685]   
28                                 WARPEHOSKI, MARTHA  [C90007626, C90007717]   
29         WEST VIRGINIA CONSERVATIVE FOUNDATION INC.  [C30001721, C90012188]   
30                                    WORKING AMERICA  [C70004981, C90011156]   
31                                 BROWN FOR CONGRESS  [C00443069, C00485391]   
32                    CALIFORNIA TEACHERS ASSOCIATION  [C70001227, C90013756]   
33                      CENTER FOR INDIVIDUAL FREEDOM  [C30001747, C90013400]   
34                              CONNOLLY FOR CONGRESS  [C00445452, C00467050]   
35                           ELECT MICHAEL WILLIAMSON  [C00440750, C00524231]   
36                     EMERGENCY COMMITTEE FOR ISRAEL  [C30001911, C90013244]   
37                            ENVIRONMENT AMERICA INC  [C70005541, C90010513]   
38                         ENVIRONMENT CALIFORNIA INC  [C70005533, C90010679]   
39  FLORIDA ALLIANCE OF PLANNED PARENTHOOD AFFILIATES  [C70004957, C90014051]   
40                                       FREEDOM PATH  [C30001986, C90012832]   
41                 FRIENDS OF ERIC SERNA FOR CONGRESS  [C00325944, C00333203]   
42                           HISPANIC LEADERSHIP FUND  [C30001879, C90013624]   
43                                  KING FOR CONGRESS  [C00373563, C00393702]   
44     LABORERS' INTERNATIONAL UNION OF NORTH AMERICA  [C70004171, C90014408]   
45                               MEDIA FUND SUPER PAC  [C00526525, C00528471]   
46                     NATIONAL EDUCATION ASSOCIATION  [C30000848, C70000492]   
47               NEW JERSEY EDUCATION ASSOCIATION PAC  [C30002133, C90012238]   
48                                PARKER FOR CONGRESS  [C00474254, C00478180]

### Multiple `cmte_id`'s for identical names &ndash; across election cycles ###

As above, we'll see how many names are mapped to different `cmte_id`'s, except now *across* election cycles

In [27]:
%%sql
with shared_cmte_name as (
    select cm.cmte_nm,
           count(distinct cm2.cmte_id) as num_cmte_ids,
           array_agg(distinct cm2.elect_cycle) as elect_cycles
      from cmte cm
      join cmte cm2 on  cm2.cmte_nm      = cm.cmte_nm
                    and cm2.cmte_id     != cm.cmte_id
                    and cm2.elect_cycle != cm.elect_cycle
     group by 1
)
select count(*) as shared_names,
       round(count(*)::numeric / :cmte_distinct_names * 100.0, 2) as pct_distinct_names
  from shared_cmte_name

 * postgresql+psycopg2://crash@caladan/fecdb
1 rows affected.


shared_names pct_distinct_names
0          1370               2.71

And now report by the level of replication (name sharing by different Committees) we have in across cycles

In [28]:
%%sql
with shared_cmte_name as (
    select cm.cmte_nm,
           count(distinct cm2.cmte_id) as num_cmte_ids,
           array_agg(distinct cm2.elect_cycle) as elect_cycles
      from cmte cm
      join cmte cm2 on  cm2.cmte_nm      = cm.cmte_nm
                    and cm2.cmte_id     != cm.cmte_id
                    and cm2.elect_cycle != cm.elect_cycle
     group by 1
)
select num_cmte_ids,
       count(*) as num_shared_names
  from shared_cmte_name
 group by 1
 order by 1 desc

 * postgresql+psycopg2://crash@caladan/fecdb
9 rows affected.


num_cmte_ids  num_shared_names
0            13                 1
1             9                 1
2             8                 1
3             7                 4
4             6                 4
5             5                21
6             4                26
7             3               147
8             2              1165

And we'll take a look at the top offenders for this replication 

In [29]:
%%sql
with shared_cmte_name as (
    select cm.cmte_nm,
           count(distinct cm2.cmte_id) as num_cmte_ids,
           array_agg(distinct cm2.cmte_id) as cmte_ids,
           count(*) as count_records
      from cmte cm
      join cmte cm2 on  cm2.cmte_nm      = cm.cmte_nm
                    and cm2.cmte_id     != cm.cmte_id
                    and cm2.elect_cycle != cm.elect_cycle
     group by 1
)
select cmte_nm,
       num_cmte_ids,
       cmte_ids
  from shared_cmte_name
 order by 2 desc, 1
 limit 50

 * postgresql+psycopg2://crash@caladan/fecdb
50 rows affected.


cmte_nm  num_cmte_ids  \
0                COLLINS FOR CONGRESS            13   
1                  KELLY FOR CONGRESS             9   
2                  MOORE FOR CONGRESS             8   
3               ANDERSON FOR CONGRESS             7   
4               CAMPBELL FOR CONGRESS             7   
5               ROBINSON FOR CONGRESS             7   
6                 ROGERS FOR CONGRESS             7   
7                KENNEDY FOR CONGRESS             6   
8                  LUCAS FOR CONGRESS             6   
9                 THOMAS FOR CONGRESS             6   
10              THOMPSON FOR CONGRESS             6   
11                 ADAMS FOR CONGRESS             5   
12            ALASKANS FOR VONDERSAAR             5   
13                 CLARK FOR CONGRESS             5   
14                  COLE FOR CONGRESS             5   
15                COOPER FOR CONGRESS             5   
16                 DAVIS FOR CONGRESS             5   
17              FLETCHER FOR CONGRESS             5   
18               GOOD GOVERNMENT FUND             5   
19                MORGAN FOR CONGRESS             5   
20                PARKER FOR CONGRESS             5   
21              PETERSON FOR CONGRESS             5   
22           PHIL NORRIS FOR CONGRESS             5   
23                PORTER FOR CONGRESS             5   
24                POWELL FOR CONGRESS             5   
25               ROBERTS FOR CONGRESS             5   
26               RUSSELL FOR CONGRESS             5   
27                 SMITH FOR CONGRESS             5   
28                    STEPP COMMITTEE             5   
29              SULLIVAN FOR CONGRESS             5   
30                WEGNER FOR CONGRESS             5   
31              WILLIAMS FOR CONGRESS             5   
32              AMERICANS FOR AMERICA             4   
33                 BROWN FOR CONGRESS             4   
34         CLEAN WATER ACTION PROJECT             4   
35  COMMITTEE TO ELECT WAYNE E HARMON             4   
36                CRAMER FOR CONGRESS             4   
37              CRAWFORD FOR CONGRESS             4   
38              FRIENDS OF DUANE SAND             4   
39                 GUILD FOR CONGRESS             4   
40           HELLER VICTORY COMMITTEE             4   
41                HUGHES FOR CONGRESS             4   
42          ISAKSON VICTORY COMMITTEE             4   
43           MARK GIBSON FOR CONGRESS             4   
44                MILLER FOR CONGRESS             4   
45               MONTOYA FOR CONGRESS             4   
46                MURRAY FOR CONGRESS             4   
47                PHELPS FOR CONGRESS             4   
48              PHILLIPS FOR CONGRESS             4   
49                 PRICE FOR CONGRESS             4   

                                             cmte_ids  
0   [C00245712, C00250605, C00270983, C00331124, C...  
1   [C00295493, C00317834, C00323493, C00352732, C...  
2   [C00320721, C00331066, C00397505, C00397927, C...  
3   [C00310235, C00338426, C00422972, C00462549, C...  
4   [C00352468, C00358416, C00398313, C00412312, C...  
5   [C00338806, C00386359, C00417212, C00428185, C...  
6   [C00343863, C00344507, C00344812, C00370155, C...  
7   [C00287078, C00347203, C00357046, C00360032, C...  
8   [C00287912, C00328922, C00366146, C00419705, C...  
9   [C00321653, C00374561, C00384222, C00649632, C...  
10  [C00356170, C00467191, C00514695, C00523365, C...  
11  [C00416388, C00467860, C00522797, C00580282, C...  
12  [C00346882, C00367599, C00389908, C00415737, C...  
13  [C00358788, C00400275, C00449991, C00452722, C...  
14  [C00337899, C00373274, C00379735, C00557108, C...  
15  [C00322032, C00376665, C00415810, C00447839, C...  
16  [C00377648, C00421909, C00439901, C00449371, C...  
17  [C00313999, C00327833, C00344150, C00364406, C...  
18  [C00340208, C00358937, C00380261, C00404004, C...  
19  [C00372730, C00463406, C00505164, C00590976, C...  
20  [C00285569, C00374074, C00449140, C00474254,

## Summary ##

### Findings ###

* Less than 1% of records have obvious flaws in the name, but there are opportunites to normalize (to make name matching more accurate, when needed)
* There are no null `cmte_id`'s and no duplicated `cmte_id`'s within an election cycle
* High-level summary of `cmte_id` longevity across two-year election cycles:
    * 28% of distinct `cmte_id`'s only appear in a single election cycle's data (i.e. 2 years)
    * 47% of distinct `cmte_id`'s appear in two or three election cycles' data (i.e. 3-6 years)
    * 25% of distinct `cmte_id`'s appear in four or more election cycles' data (i.e. 7+ years)
* 10% of Committees (identified by `cmte_id`) are listed under different names (`cmte_nm`) across election cycles
    * This represents 14% of Committees appearing in multiple election cycles
    * In some cases, it is a matter of spacing and punctuation (fixable by standard normalization), but in many other cases, the names have different wording (e.g. more or less specificity) or spelling (e.g. abbreviations)
* 2.8% of distinct Committee names in the overall data set map to 2 or more `cmte_id`'s within an election cycle
    * A similar number (actually 2.7%) of distinct Committee names are seen with different `cmte_id`'s *across* election cycles
    * Many of the same names appear in both lists of replicated/reused names

### Recommendations ###

Based on the above exploration and findings, the following enhancements to the schema are recommended:

* Normalize `cmte_nm` (and other text fields? \[see below]) during intial load (awk script)
    * Convert to uppercase
    * Collapse consecutive whitespace
    * Eliminate whitespace immediately inside of parentheses
* Create `cmte_mstr` table based on unique `cmte_id`'s (across election cycles)
    * Include `base_cmte_mstr_id` for associating Committees with different `cmte_id`'s, but identical (*or similar*) names, or otherwise deemed to be the same underlying organization

Other text fields to consider for normalization:

* `tres_nm`
* `cmte_st1`
* `cmte_st2`
* `cmte_city`
* `cmte_st`
* `connected_org_nm`